File: ios/Plugin/WatchLinkPlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(WatchLinkPlugin, "WatchLink",
           CAP_PLUGIN_METHOD(activate, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(paired, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(send, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(listen, CAPPluginReturnCallback);
           CAP_PLUGIN_METHOD(unlisten, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(hasCompanionAppInstalled, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(openPlayStoreOnWatchesWithoutApp, CAPPluginReturnPromise);
)

File: ios/Plugin/WatchLinkPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(WatchLinkPlugin)
public class WatchLinkPlugin: CAPPlugin {
    private let implementation = WatchLink()
    private var call: CAPPluginCall?

    @objc func connected(_ call: CAPPluginCall) {
        call.resolve(implementation.connected().toDict())
    }

    @objc func send(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.resolve(WatchLinkResult(ok: false, error: "Path is required").toDict())
            return
        }

        guard let message = call.getString("message") else {
            call.resolve(WatchLinkResult(ok: false, error: "Message is required").toDict())
            return
        }

        call.resolve(implementation.send(message: message, path: path).toDict())
    }

    @objc func listen(_ call: CAPPluginCall) {
        call.keepAlive = true
        self.call = call

        _ = implementation.listen() { (message: [String: Any]) -> Void in
            if let callback = self.call {
                callback.resolve(message)
            }
        }
    }

    @objc func unlisten(_ call: CAPPluginCall) {
        if let callback = self.call {
            callback.keepAlive = false
            self.call = nil;
        }

        call.resolve();
    }
}
